### PR TITLE
Harder, Better, Faster, Stronger!

### LINKED ISSUE #33: world argument
It would be really useful if there was a `/wild [player] [world]`. We use two maps, one for resources, and we are trying to allow players to /wild into the resources map

### LINKED ISSUE #42: [Request] %TIME% localization
The cooldown message uses a %TIME% variable. The time is constructed by words like `minute`. These words are not translatable like e.g:
```
  years: '&e[years] years '
  oneYear: '&e[years] year '
  day: '&e[days] days '
  oneDay: '&e[days] day '
  hour: '&e[hours] hours '
  oneHour: '&e[hours] hour '
  minutes: '&e[mins] minutes '
  oneMinute: '&e[mins] minute '
  seconds: '&e[secs] seconds '
  oneSec: '&e[secs] second '
```
or:
```
timeleft-day: "%0% day"
timeleft-days: "%0% days"
timeleft-hour: "%0% hour"
timeleft-hours: "%0% hours"
timeleft-minute: "%0% minute"
timeleft-minutes: "%0% minutes"
timeleft-second: "%0% second"
timeleft-seconds: "%0% seconds"
```
Could this be added so all text is translatable into each own language?

### PR DESCRIPTION
**Pull request by arboriginal**
Changelog:
--------------------
* perf improvements, like tell in comment above PortalzGoneWild
* another perf improvement: calc world bounds once, not each RTP
* (minor) store translated string with chat colors preprocessed,
  instead of parsing colors each time they're used.
* Add a configurable bossbar "TP in X sec"
* Add autocompletion for portals name in /wildtp delete command.
* Add coords in portals list with /wildtp list (clickable, see below)
* Add new permission + command to tp to a portal
* Add configurable potion effect when rtp (per world + per type)
* Removed "no world edit" with commands => As WE is not required
* Now you can choose if portals bypass cooldown or not (per world)
* Also per world: Can choose to freeze player while waiting for tp
  or letting him move without to cancel the tp.
* Add new dynamic permissions (for /wild [player] [world])
* Allows to tp another player to given world (resolves #33 issue)
* Fix glitch: player cancel a rtp then fall and received no damages
* fixed another potential glitch, when in delay to wait the rtp
  and drop is enabled, no fall was enabled (before to be tp),
  and depending on tp wait: can be inactive when really tp.
* Also prevent now to use /wild while falling
* Fix buggy COLLDOWN message + allows to translate time parts
  (this also resolves #42 Github issue - See note below for users)
* improvements in RandomLocationSearchTask & refactoring
  (mainly by using internal fields instead of variables
  which was calculated in several methods before).
* same for TeleportGoneWild class: refactoring / restructuring
* Merge signchange & signclick (also sign break in fact)
* Merge pre & post teleport listeners
* Make IInventoryHolder an internal class of GeeYouEye,
  as it only contain a single method, with single instruction,
  + only used by GeeYouEye. => Add an option to disable GUI too.
* Check player balance before searching random location (pay after)
* Fix directional TP (min / max were never sent to search class...)
* Allows to translate directions, and time parts (h, min, sec, ...)
* Allows to choose if tp from command / sign / portal
  can bypass delay / cooldown / cost (per world)
* Prevent to override an existing portal without forcing in command
  + add a way to cancel the process (click same block)
* Add checks when creating portals to prevent portals in portals
* Allows to set size limit for portals (per world)
* Allows to ignore portals for specific gamemodes (per world)
* Code cleaning and refactoring, also remove unused class Cashier
* Change variables encoding to UTF-1337 & write stupid comments
* Having a bug, watching the game... waza?
* Saved the cheerleader, saved the world
* Readd Herobrine

/!\ Notes for users:
--------------------
1. They can to remove COOLDOWN in Messages.yml, to let the plugin
   regenerate it, or remove the "seconds" part in their translation:
   It was buggy: you could get « you must wait 1 hour seconds ... »
   Now, they can translate hour / min / second independently
2. "NO_WE" can be removed from Messages.yml, it is no more used
   (No problem if this stays in the file.)
3. New sentences are added in the Messages.yml once the plugin
   has started, they can translate them (if needed), then reload.

### CODE REVIEW

**Comment by RoboMWM** [comment]
> another perf improvement: calc world bounds once, not each RTP

What about moving borders (growing/shrinking)

**Comment by arboriginal** [comment]
didn't think about that, as I've never used it... handled in last commit.

**Comment by RoboMWM** [comment]
ok so if I'm op it doesn't bother verifying that the location is indeed suitable or not. Either way it still attempts to tp me to the same world instead of choosing a random world (and the world I'm in is not in the randomWorlds list).

**Comment by RoboMWM** [comment]
Is this a WIP PR?

**Comment by RoboMWM** [comment]
wat is portal linking???

**Comment by RoboMWM** [comment]
Btw - ur PR is re-PRing your first PR.

Recommended that you make a new branch for each PR, or reset your branch after it's been pulled with upstream.

**Comment by RoboMWM** [comment]
Just tested https://github.com/AcmeProject/WildernessTp/pull/44/commits/767ba4ed18e381544d008ec844342ad5df8bac2 and it fixes the `/wild` command using randomWorlds, but the sign with no world specified is still trying to find a location in the same world.

Same issue with `op` status bypassing the "suitable location" check.

**Comment by arboriginal** [comment]
1. not a WIP
2. link 2 portals together => you tp from one to the other
3. this PR NOT PR-ing the previous one, this is a PR between your master and mine, I've merged before to PR, you can see the tab "Files changed" in Github to illustrate: Example, check plugin.yml, you see the version passing from 2.11 to 2.12

File: README.md [hunk 1]
```diff
@@ -35,7 +35,7 @@ A default config file will be created for you, without extra charges, into the p
 ## Hooks with other plugins
 
 * [GriefPrevention](https://www.spigotmc.org/resources/griefprevention.1884/) RTPs can take claims in consideration.
-  * Works with any other claim plugin too without any hooks!
+* Works with any other claim plugin too without any hooks!
```
Comment (RoboMWM) [inline README.md]: I purposely indented this but fair enough

Comment (arboriginal) [reply README.md]: I though you did a mistake, my bad.

**Comment by RoboMWM** [comment]
If you look in commits it's showing all the commits. Yes the version was modified later on in a different commit so GitHub is going to always show the final diff. ~~But it's also showing the diff of your prior changes.~~ hmm maybe they changed that. But the commits are still there.

Not sure what portal linking has to do with rtp but ok

**Comment by RoboMWM** [comment]
Only asked if it's WIP cuz u keep committing(?) Not sure what the status is then

**Comment by arboriginal** [comment]
Yep, I shouldn't. Now I stop pushing anything.

**Comment by RoboMWM** [comment]
Hmm didn't notice it but I see you stopped it from randomly selecting a world again for retry - initially it was this way (easy to implement before random location check), but then later made it so it did randomize this. The reasoning was to prevent constant retries with worlds filled with bad blocks (e.g. skyblock worlds) and to utilize the random "chance" set in the config more accurately.

**Comment by RoboMWM** [comment]
also the bonelessIceScream (boneless == vanilla == vanilla nether) has a ceiling. Other world generators may not. So ya that's why that exists with the check it had.

File: src/main/java/net/poweredbyhate/wildtp/WildTP.java [hunk 2]
```diff
@@ -12,54 +12,42 @@
...
+    HashMap<String, String>       aliaz;
+    HashMap<String, WorldConfig>  thugz;
     HashMap<UUID, BukkitRunnable> ohWait;
 
     public void onEnable() {
         if (Integer.valueOf(Bukkit.getBukkitVersion().split("\\.")[1]) <= 12) {
             getLogger().severe("This version of Wilderness-TP does not support your ancient server version.");
             getLogger().warning("Either update your server to 1.13, or use Wild 1.52");
-            getLogger().warning("http://r.robomwm.com/oldwild");
+            getLogger().warning("https://robomwm.com"); // URL changed because the old one no more works
```
Comment (RoboMWM) [inline src/main/java/net/poweredbyhate/wildtp/WildTP.java]: So somehow this commit reverted this commit https://github.com/AcmeProject/WildernessTp/pull/44/commits/dcf96b7decd1c1f0ffd5b7b406ecc6eb20dbb054

I'm guessing you might've not pulled changes locally and just copy-pasted your local repo over the remote repo to push this in.

**Pull request merged**